File: c/tests/04_positional_args.cpp
// SPDX-License-Identifier: MIT
// Copyright (c) 2021-2022 brainpower <brainpower at mailbox dot org>

#include "test.hpp"


TEST_CASE("positional args: general usage", "[positional]") {
  const vector<const char *> argv = {
    "/test04", "-i", "input.bin", "--", "file1", "file2",
  };

  CheckArgUPtr ca(
    checkarg_new(argv.size(), (char **)argv.data(), "test04", NULL, NULL),
    &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  int ret = checkarg_parse(ca.get());

  CHECK(ret == CA_ALLOK);

  auto args = checkarg_pos_args(ca.get());
  REQUIRE(args[0] == string("file1"));
  REQUIRE(args[1] == string("file2"));
}

TEST_CASE("positional args: args with dashes", "[positional]") {
  // these -- and - args must not be interpreted as options
  // if they appear after a --
  // when parsed as options these would trigger CA_INVOPT
  // hence failing the CHECK() below
  const char *arg1 = GENERATE("--file1", "file1", "-f", "-fg");
  const char *arg2 = GENERATE("--file2", "file2", "-g", "-hi");
  const char *arg3 = GENERATE("--file3", "file3", "-i", "-jk");

  const vector<const char *> argv = {"/test04", "-i", "input.bin", "--",
                                     arg1,      arg2, arg3};

  CheckArgUPtr ca(
    checkarg_new(argv.size(), (char **)argv.data(), "test04", NULL, NULL),
    &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);

  int ret = checkarg_parse(ca.get());

  CHECK(ret == CA_ALLOK);

  auto args = checkarg_pos_args(ca.get());
  REQUIRE(args[0] == string(arg1));
  REQUIRE(args[1] == string(arg2));
  REQUIRE(args[2] == string(arg3));
}

TEST_CASE("positional args: missing value before '--'", "[positional]") {
  // if the last option before a '--' has a value which was not given
  // the "--" shall be the value instead of the separator
  // the '-j' must be interpreted as an option
  //
  // Maybe consider an alternative here, though:
  // always have '--' be the separator
  // which would trigger a CA_MISSVAL and not interpret the '-j'
  // This is probably closer to the users intention
  const vector<const char *> argv = {
    "/test04", "-i", "--", "file1", "-j", "file2",
  };

  CheckArgUPtr ca(
    checkarg_new(argv.size(), (char **)argv.data(), "test04", NULL, NULL),
    &checkarg_free);
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);
  checkarg_add(ca.get(), 'j', "jay", "just a jay", 0, NULL);

  int ret = checkarg_parse(ca.get());

  // CHECK(ret == CA_MISSVAL); // should it be this?
  // REQUIRE(!checkarg_isset("jay"));

  CHECK(ret == CA_ALLOK);  // or this?
  REQUIRE(checkarg_value(ca.get(), "input") == string("--"));
  REQUIRE(checkarg_isset(ca.get(), "jay"));

  auto args = checkarg_pos_args(ca.get());
  REQUIRE(args[0] == string("file1"));
  REQUIRE(args[1] == string("file2"));
}

File: rust/tests/04_positional_args.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2021-2022 brainpower <brainpower at mailbox dot org>
#![feature(assert_matches)]

#[cfg(test)]
mod tests {
  use itertools::Itertools;
  use std::assert_matches::assert_matches;

  use checkarg::{CheckArg, ValueType, RC};

  #[test]
  fn general_usage() {
    let argv = vec!["/test04", "-i", "input.bin", "--", "file1", "file2"];

    let mut ca = CheckArg::new("test04");
    ca.add('i', "input", "file to read from", ValueType::Required, None);

    let rc = ca.parse(&argv);

    assert_matches!(rc, RC::Ok, "parsing failed");

    assert_eq!(
      ca.pos_args(),
      &vec!("file1".to_string(), "file2".to_string())
    );
  }

  #[test]
  fn args_with_dashes() {
    // these -- and - args must not be interpreted as options
    // if they appear after a --
    // when parsed as options these would trigger RC::InvOpt
    // hence failing the assert_matches!() below
    let all = vec![
      "--file1", "file1", "-f", "-fg", "--file2", "file2", "-g", "-hi", "--file3", "file3", "-i",
      "-jk",
    ];

    let mut ca = CheckArg::new("test04");
    ca.add('i', "input", "file to read from", ValueType::Required, None);

    let it = all.into_iter().permutations(3);
    for args in it {
      let argv = vec![
        "/test04",
        "-i",
        "input.bin",
        "--",
        args[0],
        args[1],
        args[2],
      ];
      //eprintln!("argv: {:?}", argv);
      let ret = ca.parse(&argv);

      assert_matches!(ret, RC::Ok, "parsing failed");
      assert_eq!(ca.pos_args(), &vec!(args[0], args[1], args[2]),);
    }
  }

  #[test]
  fn missing_value_before_double_dash() {
    // if the last option before a '--' has a value which was not given
    // the "--" shall be the value instead of the separator
    // the '-j' must be interpreted as an option
    //
    // Maybe consider an alternative here, though:
    // always have '--' be the separator
    // which would trigger a CA_MISSVAL and not interpret the '-j'
    // This is probably closer to the users intention

    let argv = vec!["/test04", "-i", "--", "file1", "-j", "file2"];

    let mut ca = CheckArg::new("test04");
    ca.add('i', "input", "file to read from", ValueType::Required, None);
    ca.add('j', "jay", "just a jay", ValueType::None, None);

    let rc = ca.parse(&argv);

    // should it be this?
    //assert_matches!(rc, RC::MissVal, "parsing failed");
    //assert!(!ca.isset("jay"));

    // or this?
    assert_matches!(rc, RC::Ok, "parsing failed");
    assert_eq!(ca.value("input").unwrap(), "--");
    assert!(ca.isset("jay"));

    assert_eq!(
      ca.pos_args(),
      &vec!("file1".to_string(), "file2".to_string())
    );
  }
}
